### DISCUSSION TITLE
texture_2d<f16> not allowed?

### DISCUSSION THREAD

**Discussion by AshleyScirra** [gpuweb/gpuweb#4481] [Q&A]
I was experimenting with the `shader-f16` extension in WebGPU which is now getting support in Chrome, and I noticed `texture_2d<f16>` is a compile error even when f16 is supported:

```
texture_2d<type>: type must be f32, i32 or u32
```

Indeed there are several places in the WGSL spec where it states the only types allowed are f32, i32 or u32, including in the sampling builtins like textureSample().

Is this an oversight or an intentional limitation? It seems odd and would appear to reduce the utility of f16 for saving processing time, as retrieving texture samples as f32 and then immediately having to convert that down to f16 seems to involve more processing work than if texture samples were obtained as f16 in the first place. Unless the intent is for optimizers to figure this out?

**Comment by greggman** [comment]
I'm sure someone with more knowledge in this area will respond but that seems normal to me.

If my texture is `rgba8unorm` I still use `texture_2d<f32>` not `texture_2d<unorm>` so I don't see why it would be any different for f16 textures. I also think it's not any different in other shading languages?

**Comment by kainino0x** [comment]
I've been thinking this for a long time, but I don't know if that thought was recorded anywhere.

It is theoretically desirable to be able to load texture data into f16 registers without going through an f32 intermediate. It's something that implementations can easily paper over (by implicitly going through an f32 intermediate), so if some backends could benefit from it - Metal should be able to - then I think it is a good idea. I don't know if there are other considerations that make it difficult. Will transfer this to an issue.

**Reply by AshleyScirra** [reply]
Possibly related point: you can make a fragment output with type `f16` , e.g. `@location(0) color : vec4<f16>` is allowed. However does the implementation internally use an f32 intermediate for output there?

I suppose the wider aim is to achieve true f16 support from texture sampling through to fragment output (which `texture_2d<f16>` would help with), especially as `f16` looks like a good alternative to `lowp` for fragment values in GLSL. However if `f32` conversion is involved along the way, it will probably negate the benefits of using `f16` for simple shaders (e.g. a straightforward texture fill shader that samples a texture and outputs that).

**Reply by kainino0x** [reply]
Oops forgot to link to the issue I filed, #4483

**Reply by kainino0x** [reply]
About fragment outputs, I don't know. If we emit f16 to the backends, then it's up to the drivers whether they take advantage of f16 fragment outputs or not. (Someone could check Chrome with `--enable-dawn-features=dump_shaders` which prints backend shaders to the console.) I'd guess some hardware can, while some hardware has fixed-function f32 "Output Merger" hardware.

I would be surprised if there's really a measurable impact of any f16<->f32 conversions as that should be a very straightforward operation with hardware support. Even in a trivial shader I'd expect that it'll be dwarfed by other steps in the pipeline.
